[docs] EAS Builds: add 'Advanced Credentials' page (#9704)
expo/expo: docs/pages/build/advanced-credentials.md

---

> title: Advanced Credentials

Reviewer: ```suggestion
title: Advanced credentials configuration
```

```sh
keytool \\
 -genkey -v \\
 -storetype JKS \\
 -keyalg RSA \\
 -keysize 2048 \\
 -validity 10000 \\
 -storepass KEYSTORE_PASSWORD \\
 -keypass KEY_PASSWORD \\
 -alias KEY_ALIAS \\
 -keystore release.keystore \\
 -dname "CN=com.expo.your.android.package,OU=,O=,L=,S=,C=US"
```

You can also generate it as a one of the steps in `expo eas:build --platform android`. Remember that this command does not store the newly generated keystore on your computer. If you want to take advantage of the `credentials.json` file, you'll need to download the keystore with `expo fetch:android:keystore`.

Once you have the keystore file on your computer, you should move it to the appropriate directory. We recommend you to keep your keystores in the `android/keystores` directory. **Remember to git ignore all your production keystores!** If you've run the above keytool command and placed the keystore at `android/keystores/release.keystore`, you can ignore that file by adding the following line to `.gitignore`:

```
android/keystores/release.keystore
```

> 

Reviewer: ```suggestion

Both methods for generating Keystore require `keytool` to be installed and in your PATH. If it's not available on your system you will need to install Java JDK (`keytool` is distributed with it)

```

Not sure about wording, but we should have some info on `keytool`

### Configuring credentials.json

Create `credentials.json` and configure it with the credentials:

```json
{
  "android": {
    "keystore": {
      "keystorePath": "android/keystores/release.keystore",
      "keystorePassword": "KEYSTORE_PASSWORD",
      "keyAlias": "KEY_ALIAS",
      "keyPassword": "KEY_PASSWORD"
    }
  }
}
```

> - `keystorePath` points to the keystore located on your computer. Both relative (to the project root) and absolute paths are supported.

Reviewer: > and absolute paths are supported.

We support that in code, but I never tested it with absolute path

- `keystorePassword` is the keystore password. If you've followed the previous steps it's the value of `KEYSTORE_PASSWORD`.
- `keyAlias` is the key alias. If you've followed the previous steps it's the value of `KEY_ALIAS`.
- `keyPassword` is the key password. If you've followed the previous steps it's the value of `KEY_PASSWORD`.

## iOS Credentials

### Obtaining the Distribution Certificate and Provisioning Profile

Things become more complicated when it comes to building the iOS app binary. You need a paid Apple Developer Account for starters. Next, you need to generate the Distribution Certificate and Provisioning Profile for your application. You can do that via the Apple Developer Portal or by choosing Expo to handle this for you.

If you don't know how to do it yourself, just run `expo eas:build --platform ios` and you'll be guided through the entire process. Later, you can fetch the generated credentials to your computer by running `expo fetch:ios:certs`.

> Once you have the Distribution Certiface and Provisioning Profile on your computer, you should move it to the appropriate directory. We recommend you to keep them in the `ios/certs` directory and to name them `dist.p12` and `profile.mobileprovision`, respectively. **Remember to git ignore all files in the directory!** If you've placed the credentials in the directory suggested by us, you can ignore those files by adding the following line to `.gitignore`:

Reviewer: ```suggestion
Once you have the Distribution Certificate and Provisioning Profile on your computer, you should move it to the appropriate directory. We recommend that you keep them in the `ios/certs` directory and we assume in the rest of this document they are named `dist.p12` and `profile.mobileprovision`, respectively. **Remember to git ignore all files in the directory!** If you've placed the credentials in the directory suggested by us, you can ignore those files by adding the following line to `.gitignore`:
```

{
  "android": {
    ...
  },
  "ios": {
    "provisioningProfilePath": "ios/certs/profile.mobileprovision",
    "distributionCertificate": {
      "path": "ios/certs/dist.p12",
      "password": "DISTRIBUTION_CERTIFICATE_PASSWORD"
    }
  }
}
```

- `provisioningProfilePath` points to the Provisioning Profile located on your computer. Both relative (to the project root) and absolute paths are supported.
- `distributionCertificate.path` points to the Distribution Certificate located on your computer. Both relative (to the project root) and absolute paths are supported.
- `distributionCertificate.password` is the password for the Distribution Certificate located at `distributionCertificate.path`.

## Setting Credentials Source in eas.json

### Auto Mode (Default)

Let's assume we're only building for Android and we're using the following configuration (defined in `eas.json` - [learn more about this file](../eas-json/)):

```json
{
  "builds": {
    "android": {
      "release": {
        "workflow": "generic"
      }
    }
  }
}
```

Given that configuration, `expo eas:build` resolves the credentials in the so-called **auto mode**.

The algorithm of the auto mode works like this:

> - If `credentials.json` exists and the project's credentials exist on the Expo servers:

Reviewer: ```suggestion
- If `credentials.json` entry for that platform exists and the project's credentials exist on the Expo servers:
```

  - Check if they are the same and if so - use them.
  - Otherwise, display a prompt to ask which ones should be used.

> - If `credentials.json` exists but the project's credentials do **not** exist on the Expo servers - use the credentials defined in `credentials.json`.

Reviewer: ```suggestion
- If `credentials.json` entry for that platform exists but the project's credentials do **not** exist on the Expo servers - use the credentials defined in `credentials.json`.
```

> - If `credentials.json` does **not** exist but the project's credentials exist on the Expo servers - use the credentials from the Expo servers.

Reviewer: ```suggestion
- If `credentials.json` entry for that platform does **not** exist but the project's credentials exist on the Expo servers - use the credentials from the Expo servers.
```

Before you start working on setting up your CI job, make sure you have these two files:

- `credentials.json` with paths and secrets for your credentials, something like this:

  ```json
  {
    "android": {
      "keystore": {
        "keystorePath": "android/keystores/release.keystore",
        "keystorePassword": "paofohlooZ9e",
        "keyAlias": "keyalias",
        "keyPassword": "aew1Geuthoev"
      }
    },
    "ios": {
      "provisioningProfilePath": "ios/certs/profile.mobileprovision",
      "distributionCertificate": {
        "path": "ios/certs/dist-cert.p12",
        "password": "iex3shi9Lohl"
      }
    }
  }
  ```

- `eas.json` ([learn more](../eas-json/)) with build profiles which enforce using the local `credentials.json` file:

  ```json
  {
    "builds": {
      "android": {
        "release": {
          "workflow": "generic",
          "credentialsSource": "local"
        }
      },
      "ios": {
        "release": {
          "workflow": "generic",
          "credentialsSource": "local"
        }
      }
    }
  }
  ```

> Usually, developers tend to provide CI jobs with secrets by using environment variables. One of the challenges with this apporach is that the `credentials.json` file contains a JSON object. Not all CIs allow you to set an environment variable with such value. One of the solutions to this problem is to convert the file to a base64-encoded string, set an environment variable to that value, and later decode it and restore the file on the CI.

Reviewer: >  Not all CIs allow you to set an environment variable with such value.

the environment variable is always a string so I'm not sure what you meant by that

Consider the following steps:

- Run `base64 credentials.json` in the console.
- On your CI, set the `CREDENTIALS_JSON_BASE64` environment variable with the output of the above command.

> - In the CI job, restore the file using a simple shell script:

Reviewer: ```suggestion
- In the CI job, restore the file using a simple shell command:
```
